mlir/lib/Dialect/GPU/IR/GPUDialect.cpp

```diff
@@ -1139,6 +1139,34 @@ void ShuffleOp::build(OpBuilder &builder, OperationState &result, Value value,
         mode);
 }
 
+//===----------------------------------------------------------------------===//
+// BarrierOp
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Remove gpu.barrier after gpu.barrier, the threads are already synchronized!
+struct EraseRedundantGpuBarrierOpPairs : public OpRewritePattern<BarrierOp> {
```

Comment:
There is a simpler way to write this to avoid all the boilerplate. See https://github.com/llvm/llvm-project/commit/7897a944d9e72d79f38a443afffbbfd1accfe4ad as an example.